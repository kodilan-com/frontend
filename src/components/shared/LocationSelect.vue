<script>
import { mapActions } from 'vuex';
import allLocations from '../../assets/data/locations';

export default {
  props: {
    showAll: {
      type: Boolean,
      required: false,
      default: false,
    },
    searchable: {
      type: Boolean,
      required: false,
      default: false,
    },
    value: {
      type: String,
      required: false,
      default: null,
    },
  },
  data() {
    return {
      selected: null,
      locations: null,
      options: [],
    };
  },
  methods: {
    ...mapActions(['fetchAvailableLocations']),
    handleChange(value) {
      this.$emit('input', value);
    },
    syncValue() {
      if (this.value) this.selected = this.value;
    },
    searchChange(text) {
      this.options = text.trim()
        ? this.locations.filter(l => l.toLocaleLowerCase().search(text.toLocaleLowerCase()) !== -1)
        : this.locations;
    },
  },
  watch: {
    value() {
      // Here's why we use if condition:
      // When a change occurs, the value is emit to up.
      // When the value above changes, we do not want this watcher to work unnecessarily.
      if (this.selected !== this.value) this.syncValue();
    },
  },
  created() {
    this.fetchAvailableLocations().then(() => {
      this.locations = this.showAll ? allLocations : this.$store.state.availableLocations;
      this.options = this.locations;
    });
    this.syncValue();
  },
};
</script>

<template>
  <multiselect
    :class="{ 'is-searchable': searchable }"
    v-model="selected"
    :options="options"
    @search-change="searchChange"
    :internal-search="false"
    :searchable="searchable"
    :close-on-select="true"
    :show-labels="false"
    placeholder="Şehir seçiniz..."
    @input="handleChange"
  >
    <div slot="noResult">
      Aramanızla eşleşen bir sonuç bulunamadı.
    </div>
  </multiselect>
</template>
