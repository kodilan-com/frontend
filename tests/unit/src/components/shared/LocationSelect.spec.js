import LocationSelect from '@/components/shared/LocationSelect';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Multiselect from 'vue-multiselect';
import Vuex from 'vuex';
import allLocations from '@/assets/data/locations';

const localVue = createLocalVue();

localVue.component('multiselect', Multiselect);
localVue.use(Vuex);

describe('LocationSelect.vue', () => {
  let wrapper;
  let storeStub;
  let store;
  const renderWrapper = (props) => {
    wrapper = shallowMount(LocationSelect, {
      localVue,
      store,
      propsData: {
        ...props,
      },
    });
  };

  beforeEach(() => {
    storeStub = {
      state: {
        availableLocations: ['test1', 'test2'],
      },
      actions: {
        fetchAvailableLocations: jest.fn(),
      },
    };

    store = new Vuex.Store(storeStub);

    renderWrapper();
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();

    wrapper.destroy();
  });

  it('should match with snapshot on initial', () => {
    expect(wrapper.element)
      .toMatchSnapshot();
  });

  it('should set "is-searchable" class to multiselect component when searchable prop is given as true', () => {
    renderWrapper({ searchable: true });

    expect(wrapper.find('multiselect-stub').attributes('class')).toBe('is-searchable');
  });

  it('should call fetchAvailableLocations action on initial', async () => {
    storeStub.actions.fetchAvailableLocations.mockResolvedValue({});

    renderWrapper();

    await localVue.nextTick();

    expect(storeStub.actions.fetchAvailableLocations)
      .toBeCalled();
    expect(wrapper.find('multiselect-stub')
      .attributes('options'))
      .toEqual(storeStub.state.availableLocations.join(','));
  });

  it('should show all locations if showAll is passed as true', async () => {
    storeStub.actions.fetchAvailableLocations.mockResolvedValue({});

    renderWrapper({ showAll: true });

    await localVue.nextTick();

    expect(storeStub.actions.fetchAvailableLocations)
      .toBeCalled();
    expect(wrapper.vm.options)
      .toEqual(allLocations);
    expect(wrapper.find('multiselect-stub')
      .attributes('options'))
      .toEqual(allLocations.join(','));
  });

  it('should assign given value to selected data', () => {
    renderWrapper({ value: 'newValue' });

    expect(wrapper.vm.selected).toBe('newValue');
  });

  it('should be set selected data to null on initial if value prop is not given', () => {
    expect(wrapper.vm.selected).toBe(null);
  });

  it('should emit input event', () => {
    const value = 'test';

    wrapper.vm.handleChange(value);

    expect(wrapper.emitted('input')[0][0]).toBe(value);
  });

  it('should set locations to options if given text is empty', () => {
    const text = '  ';

    wrapper.vm.searchChange(text);

    expect(wrapper.vm.options).toEqual(storeStub.state.availableLocations);
  });

  it('should show only contained locations', () => {
    const text = storeStub.state.availableLocations[0];

    wrapper.vm.searchChange(text);

    expect(wrapper.vm.options).toEqual(storeStub.state.availableLocations.slice(0, 1));
  });

  it('should call syncValue methods when value prop and selected data is different', (done) => {
    const mockMethods = {
      syncValue: jest.fn(),
    };

    renderWrapper();

    wrapper.setMethods({
      ...mockMethods,
    });
    wrapper.setData({
      selected: 'selected',
    });
    wrapper.setProps({
      value: 'changed',
    });

    expect(mockMethods.syncValue).toBeCalled();

    done();
  });

  it('should not call syncValue methods when value prop and selected data is not different', (done) => {
    const mockMethods = {
      syncValue: jest.fn(),
    };

    renderWrapper();

    wrapper.setMethods({
      ...mockMethods,
    });
    wrapper.setData({
      selected: 'changed',
    });
    wrapper.setProps({
      value: 'changed',
    });

    expect(mockMethods.syncValue).not.toBeCalled();

    done();
  });
});
