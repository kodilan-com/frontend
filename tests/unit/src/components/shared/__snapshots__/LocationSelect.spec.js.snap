// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationSelect.vue should match with snapshot on initial 1`] = `
<multiselect-stub
  allowempty="true"
  blockkeys=""
  class=""
  clearonselect="true"
  closeonselect="true"
  customlabel="function(t,e){return i(t)?\\"\\":e?t[e]:t}"
  deselectgrouplabel="Press enter to deselect group"
  deselectlabel="Press enter to remove"
  limit="99999"
  limittext="function(t){return\\"and \\".concat(t,\\" more\\")}"
  maxheight="300"
  name=""
  opendirection=""
  optionheight="40"
  options="test1,test2"
  optionslimit="1000"
  placeholder="Şehir seçiniz..."
  selectedlabel="Selected"
  selectgrouplabel="Press enter to select group"
  selectlabel="Press enter to select"
  shownooptions="true"
  shownoresults="true"
  showpointer="true"
  tabindex="0"
  tagplaceholder="Press enter to create a tag"
  tagposition="top"
>
  <div>
    
    Aramanızla eşleşen bir sonuç bulunamadı.
  
  </div>
</multiselect-stub>
`;
